src/Kupo/App/Database.hs

```diff
@@ -361,6 +368,38 @@ mkDatabase tr longestRollback bracketConnection = Database
                 ] -> yield Input{..}
             (xs :: [SQLData]) -> throwIO (UnexpectedRow addressLike [xs])
 
+    , foldInputsOutputReference = \statusFlag outputRef yield -> ReaderT $ \conn -> do
+        let matchMaybeBytes = \case
+                SQLBlob bytes -> Just bytes
+                _ -> Nothing
+        let matchMaybeWord64 = \case
+                SQLInteger (fromIntegral -> wrd) -> Just wrd
+                _ -> Nothing
```

Comment:
Since those functions are (a) re-used in two nearby places, (b) fairly generic; It'd be better to lift them out of the local lexical scope and have them top-level on this module as generic helpers.